### PR TITLE
Add color_filter

### PR DESCRIPTION
**Pull request by iory**
add rgb_color_filter
rgb_color_filter

and hsi_color_filter

hsi_color_filter

### CODE REVIEW

File: launch/hsi_color_filter.launch [hunk 1]
```diff
@@ -0,0 +1,36 @@
+<launch>
+  <arg name="node_name" default="hsi_color_filter" />
+
+  <arg name="image" default="image" doc="The image topic. Should be remapped to the name of the real image topic." />
```
Comment (mmurooka) [inline launch/hsi_color_filter.launch]: default should be `"/image_publisher/output"` to subscribe image_publisher topic in default.

**Comment by k-okada** [comment]
2016-12-09 19:09 GMT+09:00 Masaki Murooka <notifications@github.com>:

> default should be "/image_publisher/output" to subscribe image_publisher
> topic in default.
>

What does this mean? also look at
https://github.com/ros-perception/opencv_apps/blob/indigo/src/node/standalone_nodelet_exec.cpp.in#L41

--
◉ Kei Okada

**Comment by mmurooka** [comment]
> What does this mean?

In this launch file, image_publisher node is run: https://github.com/ros-perception/opencv_apps/pull/48/files#diff-e102ca2ed99024189fade3737992c127R16
So, I thought that this launch file intends to connect from image_publisher to color_filter.

I talked with @iory offline, and he said that this is not sample file so image_publisher should be removed from this launch file.

File: launch/hsi_color_filter.launch (hunk 1 above)
Comment (iory) [reply launch/hsi_color_filter.launch]: Thank you! Fixed this mistake.

Comment (mmurooka) [reply launch/hsi_color_filter.launch]: Thank you. Memo: default topic `image` is correct and image_publisher is removed from launch file.

**Comment by mmurooka** [comment]
+1 for merge this. Use case: https://github.com/mmurooka/euslib/blob/murooka-branch/demo/murooka/wholebody_manipulation_planner/object_motion_learning/launch/generate_mask_image_for_real_cube.launch#L13-L24

**Comment by k-okada** [comment]
could you update images at -> https://github.com/ros-perception/opencv_apps/pull/48#issue-184543025 so that everyone can understand the benefit of this node?

**Comment by iory** [comment]
image

**Comment by k-okada** [comment]
need better one, which we can use in wiki

--
◉ Kei Okada


2016-12-14 12:37 GMT+09:00 iory <notifications@github.com>:

**Comment by mmurooka** [comment]
Images of my use case.
before
after

**Comment by k-okada** [comment]
sorry i think original input is ok, because every one can get that rosbag
image, I just want to understand that what who created sample image want to
extract. Human face? cloth? floor????

--
◉ Kei Okada


2016-12-14 14:23 GMT+09:00 Masaki Murooka <notifications@github.com>:

**Comment by iory** [comment]
human's blue cloth.
Sample program's parameter updated.
hsi_color_filter
hsi_color_filter
rgb_color_filter
rgb_color_filter

**Comment by mmurooka** [comment]
@iory , can I see original image? It is difficult to imagine where human's blue cloth is?
Are target regions of filtering same for the examples of hsi_color_filter and rgb_color_filter? They are quite different.

**Comment by iory** [comment]
Original image is here.
src

It is difficult for me to separate colors clearly in RGB color spaces.
I think this is advantage of HSI color filter over RGB color filter.

**Comment by mmurooka** [comment]
I see. @k-okada , are images in https://github.com/ros-perception/opencv_apps/pull/48#issuecomment-267241077 good enough for wiki document? Or do you have any good idea to make good sample images? (now cloth is target for filtering. should we filter other region?)

**Comment by k-okada** [comment]
Can we show only human skins?

2016年12月20日(火) 10:38 Masaki Murooka <notifications@github.com>:

**Comment by iory** [comment]
OK.

hsi_color_filter

**Comment by k-okada** [comment]
input_color
hsi_color_filter

input_color
hsi_color_filter

input_color
hsi_color_filter

i agree this is very difficult to find good example (this means that we do not have good application for this node?), but we have to try as hard as possible.

**Comment by k-okada** [comment]
range should be 0-256, since inRange() function `lower <= src < uppder`, H shoulde be 0-360 and divide by 2?...

**Comment by k-okada** [comment]
and you also need something like
```
    if ( lower_color_range_[0] < upper_color_range_[0] ) {
      cv::inRange(hsv_image, lower_color_range_, upper_color_range_,
                  output_image);
    }else {
      cv::Scalar lower_color_range_0 = cv::Scalar(       0, s_min_, i_min_, 0);
      cv::Scalar upper_color_range_0 = cv::Scalar(h_max_/2, s_max_, i_max_, 0);
      cv::Scalar lower_color_range_360 = cv::Scalar(h_min_/2, s_min_, i_min_, 0);
      cv::Scalar upper_color_range_360 = cv::Scalar(   360/2, s_max_, i_max_, 0);
      cv::Mat output_image_0, output_image_360;
      cv::inRange(hsv_image, lower_color_range_0, upper_color_range_0, output_image_0);
      cv::inRange(hsv_image, lower_color_range_360, upper_color_range_360, output_image_360);
      output_image = output_image_0 | output_image_360;
    }
```

**Comment by iory** [comment]
Thanks to the k-okada's PR (https://github.com/iory/opencv_apps/pull/1), 
hsv and hls color filter were added into color_filter.

hsv_color filter
hsv_color_filter

hls_color filter
hls_color_filter

**Comment by mmurooka** [comment]
Cloud you merge this?

**Comment by k-okada** [comment]
@iory @mmurooka please update doc at http://wiki.ros.org/opencv_apps
note that do not forget to add `New feature as of version 1.11.15`

**Comment by iory** [comment]
OK

**Comment by iory** [comment]
Updated docs

**Comment by mmurooka** [comment]
Thank you!

**Comment by k-okada** [comment]
thank you

--
◉ Kei Okada


2017-03-07 18:32 GMT+09:00 iory <notifications@github.com>:

**Pull request merged**